src/wtforms/validators.py

```diff
@@ -609,7 +609,7 @@ class HostnameValidation(object):
     This is not a validator in and of itself, and as such is not exported.
     """
 
-    hostname_part = re.compile(r"^(xn-|[a-z0-9]+)(-[a-z0-9]+)*$", re.IGNORECASE)
+    hostname_part = re.compile(r"^(xn-|[a-z0-9_]+)(-[a-z0-9_]+)*$", re.IGNORECASE)
```

Comment:
Is a leading `_` valid, even though a leading `-` isn't?

Reply:
From some casual research it looks like leading and trailing underscores are valid for domain names [1], but underscores are not valid in hostnames [2]. I think the problem here is that the URL validator is using HostnameValidation to verify the domain name when it should be using a more permissive validator. I see that HostnameValidation is not exported so I think the intention of that validator was to be a domain name validator and to be used exclusively in URL validator. I hope that makes sense. Let me know if it would be a good idea to change the name of HostnameValidation to make the intentions of the class clearer.

[1] https://tools.ietf.org/html/rfc2181#section-11
[2] https://www.ietf.org/rfc/rfc1123.txt

Reply:
Yeah, I'm fine limiting it to domain names.